Sources/StreamVideo/Utils/AudioSession/AudioDeviceModule/AudioDeviceModule.swift

```diff
@@ -0,0 +1,294 @@
+        return 0
+    }
+
+    func audioDeviceModule(
+        _ audioDeviceModule: RTCAudioDeviceModule,
+        willEnableEngine engine: AVAudioEngine,
+        isPlayoutEnabled: Bool,
+        isRecordingEnabled: Bool
+    ) -> Int {
+        subject.send(.willEnableAudioEngine(engine))
+        isPlayingSubject.send(isPlayoutEnabled)
+        isRecordingSubject.send(isRecordingEnabled)
+        return 0
+    }
+
+    func audioDeviceModule(
+        _ audioDeviceModule: RTCAudioDeviceModule,
+        willStartEngine engine: AVAudioEngine,
+        isPlayoutEnabled: Bool,
+        isRecordingEnabled: Bool
+    ) -> Int {
+        subject.send(.willStartAudioEngine(engine))
+        isPlayingSubject.send(isPlayoutEnabled)
+        isRecordingSubject.send(isRecordingEnabled)
+        return 0
+    }
+
+    func audioDeviceModule(
+        _ audioDeviceModule: RTCAudioDeviceModule,
+        didStopEngine engine: AVAudioEngine,
+        isPlayoutEnabled: Bool,
+        isRecordingEnabled: Bool
+    ) -> Int {
+        subject.send(.didStopAudioEngine(engine))
+        audioLevelsAdapter.uninstall(on: 0)
+        isPlayingSubject.send(isPlayoutEnabled)
+        isRecordingSubject.send(isRecordingEnabled)
+        return 0
+    }
+
+    func audioDeviceModule(
+        _ audioDeviceModule: RTCAudioDeviceModule,
+        didDisableEngine engine: AVAudioEngine,
+        isPlayoutEnabled: Bool,
+        isRecordingEnabled: Bool
+    ) -> Int {
+        subject.send(.didDisableAudioEngine(engine))
+        audioLevelsAdapter.uninstall(on: 0)
+        isPlayingSubject.send(isPlayoutEnabled)
+        isRecordingSubject.send(isRecordingEnabled)
+        return 0
+    }
+
+    func audioDeviceModule(
+        _ audioDeviceModule: RTCAudioDeviceModule,
+        willReleaseEngine engine: AVAudioEngine
+    ) -> Int {
+        subject.send(.willReleaseAudioEngine(engine))
+        audioLevelsAdapter.uninstall(on: 0)
+        return 0
+    }
+
+    func audioDeviceModule(
+        _ audioDeviceModule: RTCAudioDeviceModule,
+        engine: AVAudioEngine,
+        configureInputFromSource source: AVAudioNode?,
+        toDestination destination: AVAudioNode,
+        format: AVAudioFormat,
+        context: [AnyHashable: Any]
+    ) -> Int {
+        audioLevelsAdapter.installInputTap(
+            on: destination,
+            format: format,
+            bus: 0,
+            bufferSize: 1024
+        )
+        return 0
+    }
+
+    func audioDeviceModule(
+        _ audioDeviceModule: RTCAudioDeviceModule,
+        engine: AVAudioEngine,
+        configureOutputFromSource source: AVAudioNode,
+        toDestination destination: AVAudioNode?,
+        format: AVAudioFormat,
+        context: [AnyHashable: Any]
+    ) -> Int {
+        0
+    }
+
+    func audioDeviceModuleDidUpdateDevices(
+        _ audioDeviceModule: RTCAudioDeviceModule
+    ) {
+        // TODO:
+    }
+
+    private enum CodingKeys: String, CodingKey {
```

Comment:
maybe we should put the enum at the top for consistency reasons (like the other one)?

Reply:
I generally prefer to have private members at the bottom